### PR TITLE
Fix #6564: enforce types of arguments for station name strings

### LINKED ISSUE #6564: Crash: loading save
**_Brumi_ opened the ticket and wrote:**

> The crash files are attached.
> I let a game running with AIs, and by the time I came back, it had already crashed. It did manage to create an emergency save, and loading the save causes OpenTTD to crash again.
> 
> Operating system: Windows 7 SP1 x64
> OpenTTD version: 1.7.0

**Attachments**
- [ottd-crash.7z](https://bugs.openttd.org/task/6564/getfile/10706/ottd-crash.7z) (933.29 KiB)

**Reported version**: _1.7.0_
**Operating system**: _All_

---

### PR DESCRIPTION
**Pull request by glx22**

### CODE REVIEW

**Comment by LordAro** [comment]
I guess this is a less general solution than what #7418 is (or will become) ?

**Comment by PeterN** [comment]
I think it may be a better solution than #7418 as here we know exactly what parameter types are expected. In #7418 it's a little bit vague.

**Comment by glx22** [comment]
Yes it doesn't prevent incorrect newgrf strings, but at least it doesn't crash. Similar to https://github.com/OpenTTD/OpenTTD/commit/2f042ea1988407524d86998dcba2672b23c8aaf6

**Comment by PeterN** [comment]
Can the debug message reference string ID at all, perhaps mapped to the GRF-local ID?

**Comment by glx22** [comment]
Hardly, GetInt64() knows nothing about the current string.

File: src/strings.cpp [hunk 1]
```diff
@@ -1414,8 +1417,9 @@ static char *FormatString(char *buff, const char *str_arg, StringParameters *arg
 						}
 					}
 
-					int64 args_array[] = {STR_TOWN_NAME, st->town->index, st->index};
-					StringParameters tmp_params(args_array);
+					uint64 args_array[] = {STR_TOWN_NAME, st->town->index, st->index};
+					WChar types_array[] ={0, SCC_TOWN_NAME, SCC_NUM};
```
Comment (PeterN) [inline src/strings.cpp]: Missing space, `= {`

**Review by PeterN — CHANGES_REQUESTED** [review]

**Review by PeterN — APPROVED** [review]

**Pull request merged**